src/front/glsl/builtins.rs

```diff
@@ -0,0 +1,1650 @@
+                        _ => unreachable!(),
+                    };
+
+                    declarations.push(add_builtin(&mut self.module, args, fun))
+                }
+            }
+            "dot" | "reflect" | "distance" | "ldexp" => {
+                // bits layout
+                // bit 0 - float/double
+                // bit 1 trough 2 - dims
+                for bits in 0..(0b1000) {
+                    let width = (bits & 0b1) * 4 + 4;
+                    let size = match bits >> 1 {
+                        0b00 => None,
+                        0b01 => Some(VectorSize::Bi),
+                        0b10 => Some(VectorSize::Tri),
+                        _ => Some(VectorSize::Quad),
+                    };
+                    let ty = || match size {
+                        Some(size) => TypeInner::Vector {
+                            size,
+                            kind: Sk::Float,
+                            width,
+                        },
+                        None => TypeInner::Scalar {
+                            kind: Sk::Float,
+                            width,
+                        },
+                    };
+
+                    let fun = match name.as_str() {
+                        "dot" => Builtins::MathFunction(MathFunction::Dot),
+                        "reflect" => Builtins::MathFunction(MathFunction::Reflect),
+                        "distance" => Builtins::MathFunction(MathFunction::Distance),
+                        "ldexp" => Builtins::MathFunction(MathFunction::Ldexp),
+                        _ => unreachable!(),
+                    };
+
+                    declarations.push(add_builtin(&mut self.module, vec![ty(), ty()], fun))
+                }
+            }
+            "sin" | "exp" | "exp2" | "sinh" | "cos" | "cosh" | "tan" | "tanh" | "acos" | "asin"
+            | "log" | "log2" => {
+                // bits layout
+                // bit 0 trough 1 - dims
+                for bits in 0..(0b100) {
+                    let size = match bits {
+                        0b00 => None,
+                        0b01 => Some(VectorSize::Bi),
+                        0b10 => Some(VectorSize::Tri),
+                        _ => Some(VectorSize::Quad),
+                    };
+
+                    declarations.push(add_builtin(
+                        &mut self.module,
+                        vec![match size {
+                            Some(size) => TypeInner::Vector {
+                                size,
+                                kind: Sk::Float,
+                                width: 4,
+                            },
+                            None => TypeInner::Scalar {
+                                kind: Sk::Float,
+                                width: 4,
+                            },
+                        }],
+                        Builtins::MathFunction(match name.as_str() {
+                            "sin" => MathFunction::Sin,
```

Comment:
these matches, aren't we already doing them elsewhere?

Reply:
We were doing them in `functions.rs` but I moved them here.